File: src/com/irisine/jiris/compiler/expression/IrisBinaryExpression.java
package com.irisine.jiris.compiler.expression;

import com.irisine.jiris.compiler.IrisGenerateHelper;
import org.irislang.jiris.compiler.IrisNativeJavaClass;

import com.irisine.jiris.compiler.IrisCompiler;

import net.bytebuddy.dynamic.DynamicType.Builder;
import net.bytebuddy.jar.asm.MethodVisitor;
import net.bytebuddy.jar.asm.Opcodes;

import java.util.Vector;

public class IrisBinaryExpression extends IrisExpression {

	public enum BinaryExpressionType {
		Assign,
		
		AssignAdd,
		AssignSub,
		AssignMul,
		AssignDiv,
		AssignMod,
		AssignBitAnd,
		AssignBitOr,
		AssignBitXor,
		AssignBitShr,
		AssignBitShl,
		AssignBitSar,
		AssignBitSal,

		LogicOr,
		LogicAnd,

		LogicBitOr,
		LogicBitXor,
		LogicBitAnd,

		Equal,
		NotEqual,

		GreatThan,
		GreatThanOrEqual,
		LessThan,
		LessThanOrEqual,

		BitShr,
		BitShl,
		BitSar,
		BitSal,

		Add,
		Sub,
		Mul,
		Div,
		Mod,

		Power,
	}
	
	private IrisExpression m_leftExpression = null;
	private IrisExpression m_rightExpression = null;
	private BinaryExpressionType m_type = BinaryExpressionType.Assign;
	
	public IrisBinaryExpression(BinaryExpressionType type, IrisExpression leftExpression, IrisExpression rightExpression) {
		super();
		m_leftExpression = leftExpression;
		m_rightExpression = rightExpression;
		m_type = type;
	}
	
	protected boolean OperateGenerate(String operator, IrisCompiler currentCompiler, Builder<IrisNativeJavaClass> currentBuilder, MethodVisitor visitor) {
		
		if(!m_rightExpression.Generate(currentCompiler, currentBuilder, visitor)) {
			return false;
		}

//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfResultValue());
//		visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "AddParameter", "(Lorg/irislang/jiris/core/IrisValue;)V", false);

		IrisGenerateHelper.AddParameter(visitor, currentCompiler);

		if(!m_leftExpression.Generate(currentCompiler, currentBuilder, visitor)) {
			return false;
		}

//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfResultValue());
//		visitor.visitLdcInsn(operator);
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfContextVar());
//		visitor.visitInsn(Opcodes.ICONST_1);
//		visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.getCurrentClassName(), "CallMethod",  "(Lorg/irislang/jiris/core/IrisValue;Ljava/lang/String;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;I)Lorg/irislang/jiris/core/IrisValue;", false);
//		visitor.visitVarInsn(Opcodes.ASTORE, currentCompiler.GetIndexOfResultValue());

        IrisGenerateHelper.CallMethod(visitor, currentCompiler, operator, 1, false);

//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//		visitor.visitInsn(Opcodes.ICONST_1);
//		visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "PopParameter", "(I)V", false);

        IrisGenerateHelper.PopParameter(visitor, currentCompiler, 1);

		return true;
	}
	
	protected boolean OperateAssignGenerate(String operator, IrisCompiler currentCompiler, Builder<IrisNativeJavaClass> currentBuilder, MethodVisitor visitor) {
		// calc
		if(!OperateGenerate(operator, currentCompiler, currentBuilder, visitor)) {
			return false;
		}
		
		// assign
		LeftValueResult result = m_leftExpression.LeftValue(currentCompiler, currentBuilder, visitor);
		if(!result.getResult()) {
			return false;
		}
		
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfResultValue());
//		visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "setRecord", "(Lorg/irislang/jiris/core/IrisValue;)V", false);

        IrisGenerateHelper.SetRecord(visitor, currentCompiler);
		LoadLeftValue(currentCompiler, visitor, result);
		//visitor.visitVarInsn(Opcodes.ASTORE, currentCompiler.GetIndexOfResultValue());
		
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//		visitor.visitInsn(Opcodes.ACONST_NULL);
//		visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "setRecord", "(Lorg/irislang/jiris/core/IrisValue;)V", false);

        IrisGenerateHelper.ClearRecord(visitor, currentCompiler);
		AfterLoad(currentCompiler, visitor, result);

		return true;
	}

	private void AfterLoad(IrisCompiler currentCompiler, MethodVisitor visitor, LeftValueResult result) {
		if(result.getType() == LeftValueType.MemberVariable) {
//			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//			visitor.visitInsn(Opcodes.ICONST_1);
//			visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "PopParameter", "(I)V", false);
            IrisGenerateHelper.PopParameter(visitor, currentCompiler, 1);
		} else if(result.getType() == LeftValueType.IndexVariable) {
			//visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
			//visitor.visitInsn(Opcodes.ICONST_2);
			//visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "PopParameter", "(I)V", false);
            IrisGenerateHelper.PopParameter(visitor, currentCompiler, 2);
		}
	}

	private void LoadLeftValue(IrisCompiler currentCompiler, MethodVisitor visitor, LeftValueResult result) {
		if(result.getType() != LeftValueType.MemberVariable && result.getType() != LeftValueType.IndexVariable) {
			visitor.visitLdcInsn(result.getIdentifier());

//			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//			visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "getRecord", "()Lorg/irislang/jiris/core/IrisValue;", false);

            IrisGenerateHelper.GetRecord(visitor, currentCompiler);

			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfContextVar());

			switch(result.getType()) {
			case ClassVariable:
				visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.getCurrentClassName(), "SetClassVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false);
				break;
			case Constance:
				visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.getCurrentClassName(), "SetConstance", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false);
				break;
			case GlobalVariable:
				visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.getCurrentClassName(), "SetGlobalVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false);
				break;
			case InstanceVariable:
				visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.getCurrentClassName(), "SetInstanceVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false);
				break;
			case LocalVariable:
				visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.getCurrentClassName(), "SetLocalVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false);
				break;
			default:
				break;
			}
		} else if(result.getType() == LeftValueType.MemberVariable){

		}
		else if(result.getType() == LeftValueType.IndexVariable) {
			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//			visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "getRecord", "()Lorg/irislang/jiris/core/IrisValue;", false);
            IrisGenerateHelper.GetRecord(visitor, currentCompiler);

			visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "AddParameter", "(Lorg/irislang/jiris/core/IrisValue;)V", false);

			IrisGenerateHelper.CallMethod(visitor, currentCompiler, "[]=", 2, false);
			IrisGenerateHelper.PopParameter(visitor, currentCompiler, 2);
//			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfResultValue());
//			visitor.visitLdcInsn("[]=");
//			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//			visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfContextVar());
//			visitor.visitInsn(Opcodes.ICONST_2);
//			visitor.visitMethodInsn(Opcodes.INVOKESTATIC, "org/irislang/jiris/compiler/IrisNativeJavaClass", "CallMethod", "(Lorg/irislang/jiris/core/IrisValue;Ljava/lang/String;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;I)Lorg/irislang/jiris/core/IrisValue;", false);
		}
        visitor.visitVarInsn(Opcodes.ASTORE, currentCompiler.GetIndexOfResultValue());
    }

	protected boolean AssignGenterate(IrisCompiler currentCompiler, Builder<IrisNativeJavaClass> currentBuilder, MethodVisitor visitor) {
		
		// Right value
		if(!m_rightExpression.Generate(currentCompiler, currentBuilder, visitor)) {
			return false;
		}
		
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfResultValue());
//		visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "setRecord", "(Lorg/irislang/jiris/core/IrisValue;)V", false);

        IrisGenerateHelper.SetRecord(visitor, currentCompiler);

		LeftValueResult result = m_leftExpression.LeftValue(currentCompiler, currentBuilder, visitor);
		if(!result.getResult()) {
			return false;
		}

		LoadLeftValue(currentCompiler, visitor, result);

		//visitor.visitVarInsn(Opcodes.ASTORE, currentCompiler.GetIndexOfResultValue());
		
//		visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar());
//		visitor.visitInsn(Opcodes.ACONST_NULL);
//		visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "setRecord", "(Lorg/irislang/jiris/core/IrisValue;)V", false);

        IrisGenerateHelper.ClearRecord(visitor, currentCompiler);

		AfterLoad(currentCompiler, visitor, result);

		return true;
	}
	
	
	@Override
	public boolean Generate(IrisCompiler currentCompiler, Builder<IrisNativeJavaClass> currentBuilder, MethodVisitor visitor) {
		boolean result = false;
		switch (m_type) {
		case Assign:
			result = AssignGenterate(currentCompiler, currentBuilder, visitor);
			break;
			
		case AssignAdd:
			result = OperateAssignGenerate("+", currentCompiler, currentBuilder, visitor);
			break;
		case AssignSub:
			result = OperateAssignGenerate("-", currentCompiler, currentBuilder, visitor);
			break;
		case AssignMul:
			result = OperateAssignGenerate("*", currentCompiler, currentBuilder, visitor);
			break;
		case AssignDiv:
			result = OperateAssignGenerate("/", currentCompiler, currentBuilder, visitor);
			break;
		case AssignMod:
			result = OperateAssignGenerate("%", currentCompiler, currentBuilder, visitor);
			break;
		case AssignBitAnd:
			result = OperateAssignGenerate("&", currentCompiler, currentBuilder, visitor);
			break;
		case AssignBitOr:
			result = OperateAssignGenerate("|", currentCompiler, currentBuilder, visitor);
			break;
		case AssignBitXor:
			result = OperateAssignGenerate("^", currentCompiler, currentBuilder, visitor);
			break;
		case AssignBitShr:
			result = OperateAssignGenerate(">>", currentCompiler, currentBuilder, visitor);
			break;
		case AssignBitShl:
			result = OperateAssignGenerate("<<", currentCompiler, currentBuilder, visitor);
			break;
		case AssignBitSar:
			result = OperateAssignGenerate(">>>", currentCompiler, currentBuilder, visitor);
			break;
		case AssignBitSal:
			result = OperateAssignGenerate("<<<", currentCompiler, currentBuilder, visitor);
			break;

		case LogicOr:
			result = OperateGenerate("||", currentCompiler, currentBuilder, visitor);
			break;
		case LogicAnd:
			result = OperateGenerate("&&", currentCompiler, currentBuilder, visitor);
			break;

		case LogicBitOr:
			result = OperateGenerate("|", currentCompiler, currentBuilder, visitor);
			break;
		case LogicBitXor:
			result = OperateGenerate("^", currentCompiler, currentBuilder, visitor);
			break;
		case LogicBitAnd:
			result = OperateGenerate("&", currentCompiler, currentBuilder, visitor);
			break;

		case Equal:
			result = OperateGenerate("==", currentCompiler, currentBuilder, visitor);
			break;
		case NotEqual:
			result = OperateGenerate("!=", currentCompiler, currentBuilder, visitor);
			break;

		case GreatThan:
			result = OperateGenerate(">", currentCompiler, currentBuilder, visitor);
			break;
		case GreatThanOrEqual:
			result = OperateGenerate(">=", currentCompiler, currentBuilder, visitor);
			break;
		case LessThan:
			result = OperateGenerate("<", currentCompiler, currentBuilder, visitor);
			break;
		case LessThanOrEqual:
			result = OperateGenerate("<=", currentCompiler, currentBuilder, visitor);
			break;

		case BitShr:
			result = OperateGenerate(">>", currentCompiler, currentBuilder, visitor);
			break;
		case BitShl:
			result = OperateGenerate("<<", currentCompiler, currentBuilder, visitor);
			break;
		case BitSar:
			result = OperateGenerate(">>>", currentCompiler, currentBuilder, visitor);
			break;
		case BitSal:
			result = OperateGenerate("<<<", currentCompiler, currentBuilder, visitor);
			break;

		case Add:
			result = OperateGenerate("+", currentCompiler, currentBuilder, visitor);
			break;
		case Sub:
			result = OperateGenerate("-", currentCompiler, currentBuilder, visitor);
			break;
		case Mul:
			result = OperateGenerate("*", currentCompiler, currentBuilder, visitor);
			break;
		case Div:
			result = OperateGenerate("/", currentCompiler, currentBuilder, visitor);
			break;
		case Mod:
			result = OperateGenerate("%", currentCompiler, currentBuilder, visitor);
			break;

		case Power:
			result = OperateGenerate("**", currentCompiler, currentBuilder, visitor);
			break;
		}
		return result;
	}

}

File: src/main/java/org/irislang/jiris/compiler/expression/IrisBinaryExpression.kt
package org.irislang.jiris.compiler.expression

import org.irislang.jiris.compiler.IrisGenerateHelper
import org.irislang.jiris.IrisNativeJavaClass

import org.irislang.jiris.compiler.IrisCompiler
import org.irislang.jiris.compiler.expression.IrisBinaryExpression.BinaryExpressionType.*

import net.bytebuddy.dynamic.DynamicType.Builder
import net.bytebuddy.jar.asm.MethodVisitor
import net.bytebuddy.jar.asm.Opcodes

class IrisBinaryExpression(val type: BinaryExpressionType, val leftExpression: IrisExpression, val rightExpression: IrisExpression) : IrisExpression() {

    enum class BinaryExpressionType {
        Assign,

        AssignAdd,
        AssignSub,
        AssignMul,
        AssignDiv,
        AssignMod,
        AssignBitAnd,
        AssignBitOr,
        AssignBitXor,
        AssignBitShr,
        AssignBitShl,
        AssignBitSar,
        AssignBitSal,

        LogicOr,
        LogicAnd,

        LogicBitOr,
        LogicBitXor,
        LogicBitAnd,

        Equal,
        NotEqual,

        GreatThan,
        GreatThanOrEqual,
        LessThan,
        LessThanOrEqual,

        BitShr,
        BitShl,
        BitSar,
        BitSal,

        Add,
        Sub,
        Mul,
        Div,
        Mod,

        Power
    }


    protected fun OperateGenerate(operator: String, currentCompiler: IrisCompiler, currentBuilder: Builder<IrisNativeJavaClass>, visitor: MethodVisitor): Boolean {

        if (!rightExpression.Generate(currentCompiler, currentBuilder, visitor)) {
            return false
        }

        IrisGenerateHelper.AddParameter(visitor, currentCompiler)

        if (!leftExpression.Generate(currentCompiler, currentBuilder, visitor)) {
            return false
        }

        IrisGenerateHelper.CallMethod(visitor, currentCompiler, operator, 1, false)

        IrisGenerateHelper.PopParameter(visitor, currentCompiler, 1)

        return true
    }

    protected fun OperateAssignGenerate(operator: String, currentCompiler: IrisCompiler, currentBuilder: Builder<IrisNativeJavaClass>, visitor: MethodVisitor): Boolean {
        // calc
        if (!OperateGenerate(operator, currentCompiler, currentBuilder, visitor)) {
            return false
        }

        // assign
        val result = leftExpression.LeftValue(currentCompiler, currentBuilder, visitor)
        if (!result.result) {
            return false
        }

        IrisGenerateHelper.SetRecord(visitor, currentCompiler)
        LoadLeftValue(currentCompiler, visitor, result)

        IrisGenerateHelper.ClearRecord(visitor, currentCompiler)
        AfterLoad(currentCompiler, visitor, result)

        return true
    }

    private fun AfterLoad(currentCompiler: IrisCompiler, visitor: MethodVisitor, result: IrisExpression.LeftValueResult) {
        if (result.type == IrisExpression.LeftValueType.MemberVariable) {
            IrisGenerateHelper.PopParameter(visitor, currentCompiler, 1)
        } else if (result.type == IrisExpression.LeftValueType.IndexVariable) {
            IrisGenerateHelper.PopParameter(visitor, currentCompiler, 2)
        }
    }

    private fun LoadLeftValue(currentCompiler: IrisCompiler, visitor: MethodVisitor, result: IrisExpression.LeftValueResult) {
        if (result.type != IrisExpression.LeftValueType.MemberVariable && result.type != IrisExpression.LeftValueType.IndexVariable) {
            visitor.visitLdcInsn(result.identifier)
            IrisGenerateHelper.GetRecord(visitor, currentCompiler)

            visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar())
            visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfContextVar())

            when (result.type) {
                IrisExpression.LeftValueType.ClassVariable -> visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.currentClassName, "SetClassVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false)
                IrisExpression.LeftValueType.Constance -> visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.currentClassName, "SetConstance", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false)
                IrisExpression.LeftValueType.GlobalVariable -> visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.currentClassName, "SetGlobalVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false)
                IrisExpression.LeftValueType.InstanceVariable -> visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.currentClassName, "SetInstanceVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false)
                IrisExpression.LeftValueType.LocalVariable -> visitor.visitMethodInsn(Opcodes.INVOKESTATIC, currentCompiler.currentClassName, "SetLocalVariable", "(Ljava/lang/String;Lorg/irislang/jiris/core/IrisValue;Lorg/irislang/jiris/core/IrisThreadInfo;Lorg/irislang/jiris/core/IrisContextEnvironment;)Lorg/irislang/jiris/core/IrisValue;", false)
                else -> {
                }
            }
            visitor.visitVarInsn(Opcodes.ASTORE, currentCompiler.GetIndexOfResultValue())
        } else if (result.type == IrisExpression.LeftValueType.MemberVariable) {
            visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar())
            IrisGenerateHelper.GetRecord(visitor, currentCompiler)
            visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo",
                    "AddParameter", "(Lorg/irislang/jiris/core/IrisValue;)V", false)

            IrisGenerateHelper.CallMethod(visitor, currentCompiler, result.identifier, 1, false)
        } else if (result.type == IrisExpression.LeftValueType.IndexVariable) {
            visitor.visitVarInsn(Opcodes.ALOAD, currentCompiler.GetIndexOfThreadInfoVar())
            IrisGenerateHelper.GetRecord(visitor, currentCompiler)

            visitor.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "org/irislang/jiris/core/IrisThreadInfo", "AddParameter", "(Lorg/irislang/jiris/core/IrisValue;)V", false)

            IrisGenerateHelper.CallMethod(visitor, currentCompiler, "[]=", 2, false)
        }
    }

    protected fun AssignGenterate(currentCompiler: IrisCompiler, currentBuilder: Builder<IrisNativeJavaClass>, visitor: MethodVisitor): Boolean {

        // Right value
        if (!rightExpression.Generate(currentCompiler, currentBuilder, visitor)) {
            return false
        }

        IrisGenerateHelper.SetRecord(visitor, currentCompiler)

        val result = leftExpression.LeftValue(currentCompiler, currentBuilder, visitor)
        if (!result.result) {
            return false
        }

        LoadLeftValue(currentCompiler, visitor, result)

        IrisGenerateHelper.ClearRecord(visitor, currentCompiler)

        AfterLoad(currentCompiler, visitor, result)

        return true
    }


    override fun Generate(currentCompiler: IrisCompiler, currentBuilder: Builder<IrisNativeJavaClass>, visitor: MethodVisitor): Boolean {
        IrisGenerateHelper.SetLineNumber(visitor, currentCompiler, lineNumber)

        var result = false
        when (type) {
            Assign -> result = AssignGenterate(currentCompiler, currentBuilder, visitor)

            AssignAdd -> result = OperateAssignGenerate("+", currentCompiler, currentBuilder, visitor)
            AssignSub -> result = OperateAssignGenerate("-", currentCompiler, currentBuilder, visitor)
            AssignMul -> result = OperateAssignGenerate("*", currentCompiler, currentBuilder, visitor)
            AssignDiv -> result = OperateAssignGenerate("/", currentCompiler, currentBuilder, visitor)
            AssignMod -> result = OperateAssignGenerate("%", currentCompiler, currentBuilder, visitor)
            AssignBitAnd -> result = OperateAssignGenerate("&", currentCompiler, currentBuilder, visitor)
            AssignBitOr -> result = OperateAssignGenerate("|", currentCompiler, currentBuilder, visitor)
            AssignBitXor -> result = OperateAssignGenerate("^", currentCompiler, currentBuilder, visitor)
            AssignBitShr -> result = OperateAssignGenerate(">>", currentCompiler, currentBuilder, visitor)
            AssignBitShl -> result = OperateAssignGenerate("<<", currentCompiler, currentBuilder, visitor)
            AssignBitSar -> result = OperateAssignGenerate(">>>", currentCompiler, currentBuilder, visitor)
            AssignBitSal -> result = OperateAssignGenerate("<<<", currentCompiler, currentBuilder, visitor)

            LogicOr -> result = OperateGenerate("||", currentCompiler, currentBuilder, visitor)
            LogicAnd -> result = OperateGenerate("&&", currentCompiler, currentBuilder, visitor)

            LogicBitOr -> result = OperateGenerate("|", currentCompiler, currentBuilder, visitor)
            LogicBitXor -> result = OperateGenerate("^", currentCompiler, currentBuilder, visitor)
            LogicBitAnd -> result = OperateGenerate("&", currentCompiler, currentBuilder, visitor)

            Equal -> result = OperateGenerate("==", currentCompiler, currentBuilder, visitor)
            NotEqual -> result = OperateGenerate("!=", currentCompiler, currentBuilder, visitor)

            GreatThan -> result = OperateGenerate(">", currentCompiler, currentBuilder, visitor)
            GreatThanOrEqual -> result = OperateGenerate(">=", currentCompiler, currentBuilder, visitor)
            LessThan -> result = OperateGenerate("<", currentCompiler, currentBuilder, visitor)
            LessThanOrEqual -> result = OperateGenerate("<=", currentCompiler, currentBuilder, visitor)

            BitShr -> result = OperateGenerate(">>", currentCompiler, currentBuilder, visitor)
            BitShl -> result = OperateGenerate("<<", currentCompiler, currentBuilder, visitor)
            BitSar -> result = OperateGenerate(">>>", currentCompiler, currentBuilder, visitor)
            BitSal -> result = OperateGenerate("<<<", currentCompiler, currentBuilder, visitor)

            Add -> result = OperateGenerate("+", currentCompiler, currentBuilder, visitor)
            Sub -> result = OperateGenerate("-", currentCompiler, currentBuilder, visitor)
            Mul -> result = OperateGenerate("*", currentCompiler, currentBuilder, visitor)
            Div -> result = OperateGenerate("/", currentCompiler, currentBuilder, visitor)
            Mod -> result = OperateGenerate("%", currentCompiler, currentBuilder, visitor)

            Power -> result = OperateGenerate("**", currentCompiler, currentBuilder, visitor)
        }
        return result
    }

}
